### PR TITLE
Handle TKR upgrade in ClusterBootstrap controller

### LINKED ISSUE #1584: Handle TKR upgrade for TanzuClusterBootStrap
**Describe the feature request**

When TKR changes for a cluster and is different from resolvedTKR in the status of TanzuClusterBootstrap, it should trigger an upgrade of packages. See placeholder TODO in vmware-tanzu/tanzu-framework#1541

**Affected product area (please put an X in all that apply)**

- [ ] APIs
- [x ] Addons
- [ ] CLI
- [ ] Docs
- [ ] IAM
- [ ] Installation
- [ ] Plugin
- [ ] Security
- [ ] Test and Release
- [ ] User Experience

**Additional context**

### PR DESCRIPTION
**Pull request by blc1996**
Signed-off-by: Lucheng Bao <luchengb@vmware.com>

### What this PR does / why we need it
Handle TKR upgrade for TanzuClusterBootStrap 

Upgrade the refName of all the core packages
Package updates keep the users' customization in valuesFrom
We assume:
1. ClusterBootstrapTemplate will always have Kapp and CNI package available
2. ClusterBootstrapTemplate will always have consistent core packages refNames in different TKR versions (same name, different version)
3. The Group and Kind for default core package providers will not change across different TKR versions
4. All packages, including additional packages, can't be deleted (meaning the package refName can't be changed, only allow version bump)
5. We will keep users' customization on valuesFrom of each package, users are responsible for the correctness of the content they put in will work with the next version.

### Which issue(s) this PR fixes

Fixes #1584

### Describe testing done for PR

Integration test for TKR upgrade is added

### Release note

```release-note
package-based-lcm: Handle TKR upgrade in ClusterBootStrap controller
```

### PR Checklist

- [x] Squash the commits into one or a small number of logical commits
      
- [x] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [x] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

File: addons/controllers/clusterbootstrap_controller.go [hunk 1]
```diff
@@ -785,21 +913,30 @@ func patchSecretWithLabels(secret *corev1.Secret, pkgName, clusterName string) b
...
-	var createdSecrets []*corev1.Secret
-
 	packages := append([]*runtanzuv1alpha3.ClusterBootstrapPackage{
 		bootstrap.Spec.CPI,
 		bootstrap.Spec.CSI,
 		bootstrap.Spec.Kapp,
 	}, bootstrap.Spec.CNIs...)
 	packages = append(packages, bootstrap.Spec.AdditionalPackages...)
 
+	return r.cloneSecretsAndProvidersFromList(cluster, packages, templateNS, log)
```
Comment (maralavi) [inline addons/controllers/clusterbootstrap_controller.go:923]: Could L320 get refactored to add the following and then also call ```cloneSecretsAndProvidersFromList()``` at line 320? It is a bit confusing to have both ```cloneSecretsAndProviders``` and ```cloneSecretsAndProvidersFromList```, and with this change ```cloneSecretsAndProviders``` would be very small function and doesn't need to be a separate function of its own
```

	packages := append([]*runtanzuv1alpha3.ClusterBootstrapPackage{
		bootstrap.Spec.CPI,
		bootstrap.Spec.CSI,
		bootstrap.Spec.Kapp,
	}, bootstrap.Spec.CNIs...)
	packages = append(packages, bootstrap.Spec.AdditionalPackages...)
```

Comment (blc1996) [reply addons/controllers/clusterbootstrap_controller.go:923]: Sounds like a plan

Comment (vijaykatam) [reply addons/controllers/clusterbootstrap_controller.go:923]: +1 combine both and caller explicitly provides the packages whose secrets and providers need to be cloned.

File: addons/controllers/clusterbootstrap_controller.go [hunk 2]
```diff
@@ -949,7 +1095,7 @@ func (r *ClusterBootstrapReconciler) updateValuesFromSecret(cluster *clusterapiv
 		newSecret = secret.DeepCopy()
 		newSecret.ObjectMeta.Reset()
 		newSecret.Name = fmt.Sprintf("%s-%s-package", cluster.Name, pkgRefName)
-		newSecret.Namespace = bootstrap.Namespace
```
Comment (vijaykatam) [inline addons/controllers/clusterbootstrap_controller.go:952]: Curious why this change is being made. While this works, it is technically namespace of ClusterBootstrap.

**Review by vijaykatam — COMMENTED** [review]
Few things, looks like we need more issues for upgrade work but I was expecting to handle merge of new configs in this PR. If not lets get another issue logged. It is lower priority for MVP so lets not spend cycles now.

For MVP, we need to handle TKGS upgrade case. Lets sync up next week to log appropriate issues
1. set pause cluster to false after rolling out package installs
2. Add tkr node labels for packageinstalls to target specific VMs
3. Handle deployment and daemonset upgrade.

File: addons/controllers/clusterbootstrap_controller.go [hunk 3]
```diff
@@ -287,68 +287,196 @@ func (r *ClusterBootstrapReconciler) createOrPatchClusterBootstrapFromTemplate(c
...
+	}
+
+	// Upgrade the refName of all the core packages
+	// Package updates keep the users' customization in valuesFrom
+	// We assume:
+	//    1. ClusterBootstrapTemplate will always have Kapp and CNI package available
+	//    2. ClusterBootstrapTemplate will always have consistent core packages refNames in different TKR versions (same name, different version)
+	//    3. The Group and Kind for default core package providers will not change across different TKR versions
+	//    4. All packages, including additional packages, can't be deleted (meaning the package refName can't be changed, only allow version bump)
+	//    5. We will keep users' customization on valuesFrom of each package, users are responsible for the correctness of the content they put in will work with the next version.
```
[the remark this answers is not in the document]
Comment (blc1996) [reply addons/controllers/clusterbootstrap_controller.go:382]: This MR assumes `The Group and Kind for default core package providers will not change across different TKR versions`, and the newly introduced provider version can be handled by conversion webhook. 

Let's log another issue to keep track of the config merging.

File: addons/controllers/clusterbootstrap_controller.go [hunk 4]
```diff
@@ -287,68 +287,196 @@ func (r *ClusterBootstrapReconciler) createOrPatchClusterBootstrapFromTemplate(c
...
+	// CSI and CPI can be nil, only update if it's present
+	// According to assumption 2, no need to do nil check on template
+	if updatedClusterBootstrap.Spec.CSI == nil {
+		updatedClusterBootstrap.Spec.CSI = clusterBootstrapTemplate.Spec.CSI.DeepCopy()
+	} else {
+		updatedClusterBootstrap.Spec.CSI.RefName = clusterBootstrapTemplate.Spec.CSI.RefName
+	}
+
+	if updatedClusterBootstrap.Spec.CPI == nil {
+		updatedClusterBootstrap.Spec.CSI = clusterBootstrapTemplate.Spec.CSI.DeepCopy()
```
[the remark this answers is not in the document]
Comment (blc1996) [reply addons/controllers/clusterbootstrap_controller.go:413]: Nica catch!

File: addons/controllers/clusterbootstrap_controller.go (hunk 2 above)
Comment (blc1996) [reply addons/controllers/clusterbootstrap_controller.go:952]: Because I wanted to minimize the parameters passed into each function. Previuosly, the only purpose of passing in bootstrap is to use its namespace. While the namespace of it will always be the same as cluster.namespace, so I just removed it.
```
clusterBootstrap := &runtanzuv1alpha3.ClusterBootstrap{}
err := r.Client.Get(r.context, client.ObjectKeyFromObject(cluster), clusterBootstrap)
```

File: addons/controllers/clusterbootstrap_controller.go [hunk 5]
```diff
@@ -287,68 +287,196 @@ func (r *ClusterBootstrapReconciler) createOrPatchClusterBootstrapFromTemplate(c
...
+	if err != nil {
+		return nil, err
+	}
+	if clusterBootstrapTemplate.Spec == nil || updatedClusterBootstrap.Spec == nil {
+		return nil, errors.New("ClusterBootstrap and ClusterBootstrapTemplate spec can't be nil")
+	}
+
+	// Upgrade the refName of all the core packages
+	// Package updates keep the users' customization in valuesFrom
+	// We assume:
```
[the remark this answers is not in the document]
Comment (blc1996) [reply addons/controllers/clusterbootstrap_controller.go]: Comment added

File: addons/controllers/clusterbootstrap_controller.go [hunk 6]
```diff
@@ -287,68 +287,196 @@ func (r *ClusterBootstrapReconciler) createOrPatchClusterBootstrapFromTemplate(c
...
+
+	// Since we don't allow users to delete additional packages in our webhook
+	// Meaning the users will not be able to customize the packageRefName
+	// Find all the corresponding pairs in ClusterBootstrap and new ClusterBootstrapTemplate to update
+	// Add the additional package if it's only present in the new ClusterBootstrapTemplate
+	// Leave the package as it is if it's only present in ClusterBootstrap but not in the new Template
+	additionalPackageMap := map[string]*runtanzuv1alpha3.ClusterBootstrapPackage{}
+
+	for _, pkg := range updatedClusterBootstrap.Spec.AdditionalPackages {
+		packageRefName, _, err := util.GetPackageMetadata(r.context, r.liveClient, pkg.RefName, cluster.Namespace)
```
[the remark this answers is not in the document]
Comment (blc1996) [reply addons/controllers/clusterbootstrap_controller.go:439]: I saw Daniel was using packageRefName so I followed suit. I think the reason here is the field under package CR is also called refName, so it's named packageRefName, which the `RefName = packageRefName + version`.

Should I change it in all the places?

File: addons/controllers/clusterbootstrap_controller.go [hunk 7]
```diff
@@ -287,68 +287,196 @@ func (r *ClusterBootstrapReconciler) createOrPatchClusterBootstrapFromTemplate(c
...
+	} else {
+		updatedClusterBootstrap.Spec.CSI.RefName = clusterBootstrapTemplate.Spec.CSI.RefName
+	}
+
+	if updatedClusterBootstrap.Spec.CPI == nil {
+		updatedClusterBootstrap.Spec.CSI = clusterBootstrapTemplate.Spec.CSI.DeepCopy()
+	} else {
+		updatedClusterBootstrap.Spec.CSI.RefName = clusterBootstrapTemplate.Spec.CSI.RefName
+	}
+
```
[the remark this answers is not in the document]
Comment (blc1996) [reply addons/controllers/clusterbootstrap_controller.go:430]: Added

File: addons/controllers/clusterbootstrap_controller.go [hunk 8]
```diff
@@ -287,68 +287,196 @@ func (r *ClusterBootstrapReconciler) createOrPatchClusterBootstrapFromTemplate(c
...
+	//    2. ClusterBootstrapTemplate will always have consistent core packages refNames in different TKR versions (same name, different version)
+	//    3. The Group and Kind for default core package providers will not change across different TKR versions
+	//    4. All packages, including additional packages, can't be deleted (meaning the package refName can't be changed, only allow version bump)
+	//    5. We will keep users' customization on valuesFrom of each package, users are responsible for the correctness of the content they put in will work with the next version.
+	newCNIPkg, err := r.getCNIForClusterBootstrap(clusterBootstrapTemplate, cluster, log)
+	if err != nil {
+		log.Error(err, "unable to get the selected CNI")
+		return nil, err
+	}
+	if len(updatedClusterBootstrap.Spec.CNIs) != 1 {
```
[the remark this answers is not in the document]
Comment (blc1996) [reply addons/controllers/clusterbootstrap_controller.go:397]: Will rebase after Marjan has her fix in.

File: addons/controllers/clusterbootstrap_controller.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (blc1996) [reply addons/controllers/clusterbootstrap_controller.go:391]: If user decide to use secretRef/Inline, they are on their own. We'll update the refName anyways, buy they should take care of the configs by themself.

File: addons/controllers/clusterbootstrap_controller.go [hunk 9]
```diff
@@ -284,62 +284,197 @@ func (r *ClusterBootstrapReconciler) createOrPatchClusterBootstrapFromTemplate(c
 
 	if clusterBootstrap.UID == "" {
 		log.Info("ClusterBootstrap for cluster does not exist, cloning from template")
+		return r.createClusterBootstrapFromTemplate(cluster, clusterBootstrapTemplate, tkrName, log)
+	}
+	// TODO: Add a warning prompt in CLI before cluster upgrade. Telling the user to make sure the compatibility if there is any customization in valuesFrom of packages
```
Comment (vijaykatam) [inline addons/controllers/clusterbootstrap_controller.go:289]: nit: cleanup commentary around CLI warning prompt

Comment (vijaykatam) [inline addons/controllers/clusterbootstrap_controller.go]: nit: cleanup commentary around CLI warning prompt

File: addons/controllers/clusterbootstrap_controller.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (vijaykatam) [reply addons/controllers/clusterbootstrap_controller.go:425]: not a blocker but maybe consider in another change

[the remark this answers is not in the document]
Comment (vijaykatam) [reply addons/controllers/clusterbootstrap_controller.go:425]: not a blocker but maybe consider in another change

**Review by vijaykatam — DISMISSED** [review]

File: addons/controllers/clusterbootstrap_controller.go [hunk 10]
```diff
@@ -284,62 +284,195 @@ func (r *ClusterBootstrapReconciler) createOrPatchClusterBootstrapFromTemplate(c
...
+		r.Log.Error(err, fmt.Sprintf("unable to update the status of ClusterBootstrap %s/%s", clusterBootstrap.Namespace, clusterBootstrap.Name))
+		return nil, err
+	}
+	r.Log.Info("cloned clusterBootstrap", "clusterBootstrap", clusterBootstrap)
+
+	return clusterBootstrap, nil
+}
+
+// patchClusterBootstrapFromTemplate will patch ClusterBootstrap associated with a cluster in case of TKR upgrade
+func (r *ClusterBootstrapReconciler) patchClusterBootstrapFromTemplate( // nolint:funlen,gocyclo
```
Comment (vijaykatam) [inline addons/controllers/clusterbootstrap_controller.go:354]: Just noticed these, please refactor to remove lint suppression.

**Review by vijaykatam — COMMENTED** [review]
Remove lint suppress and should be good to merge.

File: addons/controllers/clusterbootstrap_controller.go (hunk 10 above)
Comment (blc1996) [reply addons/controllers/clusterbootstrap_controller.go:354]: ack

Comment (blc1996) [reply addons/controllers/clusterbootstrap_controller.go]: Hi @vijaykatam , my recent commit have addressed the lint issue

File: addons/controllers/clusterbootstrap_controller.go (hunk 9 above)
Comment (blc1996) [reply addons/controllers/clusterbootstrap_controller.go]: removed

**Review by vijaykatam — APPROVED** [review]
